### PR TITLE
Fallback configuration for alertmanager

### PR DESCRIPTION
**Pull request by bboreham**
When the config for an instance is unusable (generally because it is blank), allow the alertmanager to be configured with a fallback.

We also allow the instance-id to be encoded in the Slack URL, triggered by a magic value, because multitenant.

This addresses #529, in that you no longer get a continuous stream of errors about having no valid config.

### CODE REVIEW

**Comment by bboreham** [comment]
I got round to trying this; it's still very noisy in the logs because it's trying to hit a webhook that doesn't exist.

Some other ideas:
 * implement a sink for webhooks inside the alertmanager process, so they are silently ignored
 * add a 'null' receiver to prometheus alertmanager
 * have the ruler examine the route config, so it can tell it's invalid
 * have the multitenant alertmanager return 204 instead of 404 when it doesn't have a config (from @aaron7 )

File: pkg/alertmanager/multitenant.go [hunk 1]
```diff
@@ -364,18 +380,36 @@ func (am *MultitenantAlertmanager) addNewConfigs(cfgs map[string]configs.View) {
...
+			return fmt.Errorf("invalid Cortex configuration for %v: %v", userID, err)
+		}
+		if am.fallbackConfig != nil {
+			log.Infof("invalid Cortex configuration; using fallback for %v", userID)
+			amConfig = am.fallbackConfig
+		}
+	}
+
+	// Magic ability to configure a Slack receiver if config requests it
+	if am.cfg.FallbackSlackRoot != "" {
```
Comment (jml) [inline pkg/alertmanager/multitenant.go]: Why is this called `FallbackSlackRoot`, given that it's applied to all configs, not just fallback ones?

File: pkg/alertmanager/multitenant.go [hunk 2]
```diff
@@ -364,18 +380,36 @@ func (am *MultitenantAlertmanager) addNewConfigs(cfgs map[string]configs.View) {
...
+		for _, r := range amConfig.Receivers {
+			for _, s := range r.SlackConfigs {
+				if s.APIURL == "internal://monitor" {
```
Comment (jml) [inline pkg/alertmanager/multitenant.go]: Maybe move this to a constant with a comment that explains what it's for & why we need it?

File: pkg/alertmanager/multitenant.go [hunk 3]
```diff
@@ -364,18 +380,36 @@ func (am *MultitenantAlertmanager) addNewConfigs(cfgs map[string]configs.View) {
...
+					s.APIURL = amconfig.Secret(am.cfg.FallbackSlackRoot + "/" + userID + "/monitor")
+				}
+			}
+		}
```
Comment (jml) [inline pkg/alertmanager/multitenant.go]: This function is getting big enough that I'd consider extracting out the bit that turns `configs.Config` into whatever `amConfig` is. Up to you though.

File: pkg/alertmanager/multitenant.go (hunk 1 above)
Comment (bboreham) [reply pkg/alertmanager/multitenant.go]: `MagicSlackRoot` ?

Comment (jml) [reply pkg/alertmanager/multitenant.go]: 👍

**Comment by bboreham** [comment]
Updated in view of review comments; PTAL

File: pkg/alertmanager/multitenant.go [hunk 4]
```diff
@@ -361,11 +381,34 @@ func (am *MultitenantAlertmanager) addNewConfigs(cfgs map[string]configs.View) {
 	totalConfigs.Set(float64(len(am.cfgs)))
 }
 
+func (am *MultitenantAlertmanager) transformConfig(userID string, amConfig **amconfig.Config) error {
```
Comment (jml) [inline pkg/alertmanager/multitenant.go]: I'd return the new one rather than use a double pointer but YMMV.

**Comment by bboreham** [comment]
I changed it to return a new object, and thus spotted it was modifying the underlying data, which made it less multitenant than we might have hoped.

**Comment by shaarmaayushh** [comment]
i am still having this same issue
ts=2025-08-27T08:25:30.409100501Z caller=notifier.go:544 level=error user=fake alertmanager=http://cortex-mon-cluster-alertmanager:8080/api/v2/alerts count=1 msg="Error sending alert" err="bad response status 404 Not Found"

can someone tell me how to resolve this man

**Pull request merged**